docs/05-rfcs/2023-01-21-wing-plugins.md

```diff
@@ -0,0 +1,258 @@
+
+  **Plugin Example:**
+  ```js
+  class IamNamingConventionsAspect {
+    constructor() {
+      this.namePrefix = 'SOME_PREFIX';
+      this.permissionsBoundaryArn = 'SOME_ARN';
+    }
+    visit(node) {
+      if (node instanceof IamRole) {
+        node.namePrefix = this.namePrefix;
+        node.permissionsBoundary = this.permissionsBoundaryArn;
+      }
+    }
+  }
+  class IamNamingConventions {
+    constructor() { }
+    preSynthesis(app) {
+      Aspects.of(app).add(new IamNamingConventionsAspect());
+    }
+  }
+  ```
+
+- An organization requires its cloud resources to follow strict security requirements to maintain their HIPPA compliance. These rules include encryption at rest, versioning, replication, isolation, and more. These rules can be encapsulated in a plugin that can be applied to all resources.
+  **Plugin Example: TODO**
+
+- An application team is provided a VPC in their account that was provisioned by the platform engineering team. It is required that all the resources in the application are deployed into this VPC.
+  **Plugin Example:**
+  ```js
+  class SandboxVPCAspect {
+    constructor() { }
+    visit(node) {
+      if (node instanceof LambdaFunction) {
+        node.putVpcConfig({
+          subnetIds: ["SOME_SUBNET_ID"],
+          securityGroupIds: ["SOME_SG_ID"]
+        });
+      }
+    }
+  }
+  class SandboxVPCPlugin {
+    constructor() { }
+    preSynthesis(app) {
+      Aspects.of(app).add(new SandboxVPCAspect());
+    }
+  }
+  ```
+
+- Security policies in an organization require the use of customer managed keys for all resources that support encryption at rest. 
+  **Plugin Example: TODO**
+
+- To better manage and understand costs, an organization requires that all resources are tagged with a certain set of tags.
+  **Plugin Example: TODO**
+
+- An organization's security team has decided that all IAM policies should not contain Allow "*" actions. A plugin can be written to scan the synthesized config and throw an error if a policy statement is found, and where it is found.
+  **Plugin Example:**
+  ```js
+  class ForbidWildcards {
+    constructor() { }
+    validate(config) {
+      const resources = config.resource;
+      if (resources.aws_iam_role_policy) {
+        const policyKeys = Object.keys(resources.aws_iam_role_policy);
+        for (const key of policyKeys) {
+          const policy = resources.aws_iam_role_policy[key].policy;
+          const statements = JSON.parse(policy).Statement;
+          for (const statement of statements) {
+            const actions = statement.Action;
+            if (actions.includes("*") && statement.Effect === "Allow") {
+              throw new Error(`Found a wildcard allow action in policy: ${key}`);
```

Comment:
Opinion: Policy as code and object mutation (what I thought the plugin was) are two different use cases.

Reply:
hmm, I had not thought of them as being very different. I think that the former implies the latter. Where policies are implemented via mutations (in most cases)

Maybe room for discussions and clarifications, would be happy to hear more 😃